### DISCUSSION TITLE
Optimizing finetuning "en_core_web_trf" for sentiment classification

### DISCUSSION THREAD

**Discussion by traceymai** [explosion/spaCy#7081] [Help: Coding & Implementations]
Hey guys, I've been trying to write a Python script to fine-tune the transformers based "en_core_web_trf" model for sentiment classification. Our train dataset is relatively small, consisting of around 15000 sentences. However after the model is trained, to my surprise, evaluation shows subpar performance compared to v2's "en_core_web_lg", so I'm just wondering if there's anything I can do to improve this or if I've wrongly implemented anything in the training loop? My code is attached below. Before I was looking at optimising through using a learning rate scheduler and decreasing the number of epochs, but the performance still stays at about the same accuracy. Any help is much appreciated. I'm training the model on 10 iterations.

Format of training_data: `[("Hi my name is Adam",
                      {"cats": {"POSITIVE": 0, "NEGATIVE": 0, "NEUTRAL":1}}),
                     ("What an amazing view! I can't help but loving it",
                      {"cats": {"POSITIVE": 1, "NEGATIVE": 0, "NEUTRAL":0}}),
                     ("Ugh he makes me tired all the time",
                      {"cats": {"POSITIVE": 0, "NEGATIVE": 1, "NEUTRAL":0}})]`

```
  def get_examples(_training_data, _nlp_model):
      _examples_iter = []
      for _instance in _training_data:
          _doc = _nlp_model.make_doc(_instance[0])
          _gold_dict = _instance[1]
          _example = Example.from_dict(_doc, _gold_dict)
          _examples_iter.append(_example)
          # return [example]
      return _examples_iter
  
  
  def lr_scheduler():
      lr = [0.0001, 0.0005, 0.001, 0.005, 0.01]
      while True:
          for i in lr:
              yield i
          for u in reversed(lr):
              yield u
  
  def convert_training_batch_to_examples_iterable(_batch, _nlp_model):
      _examples_iterable = []
      for _instance in _batch:
          _doc = _nlp_model.make_doc(_instance[0])
          _gold_dict = _instance[1]
          _example = Example.from_dict(_doc, _gold_dict)
          _examples_iterable.append(_example)
      return _examples_iterable`
  
  
with _nlp_model.select_pipes(enable=["textcat"]):  
    _optimizer = _nlp_model.resume_training()
    _optimizer.schedules["learn_rate"] = lr_scheduler()
     def get_examples_func():
           return get_examples(_training_data=_training_data, _nlp_model=_nlp_model)
     _textClassifier.initialize(get_examples_func, nlp=_nlp_model)
     print("Training the model...")
     print('{:^9}\t{:^9}\t{:^9}\t{:^9}\t{:^9}'.format('<<--------------------------->>', 'LOSS(↓)', 'Precision(↑)',
                                                           'Recall(↑)', 'F-score(↑)'))
      # Performing training
      for i in tqdm(range(_num_training_iter), position=0, leave=True, ncols=30):
          _optimizer.step_schedules()
          _losses = {}
          _training_batches = minibatch(_training_data,
                               size=compounding(4., 32., 1.001)) # batch up training examples using spaCy's minibatch
          for _batch_train in _training_batches:
              _examples_iterable = convert_training_batch_to_examples_iterable(_batch = _batch_train, _nlp_model=_nlp_model)
              _losses = _nlp_model.update(_examples_iterable, sgd=_optimizer, drop=0.2, losses=_losses)
          # Calling the evaluate() function and printing the scores
          _testText, _testCats = _test_data[0], _test_data[1]
          with _textClassifier.model.use_params(_optimizer.averages):
              # evaluate on the test _data split off from training _data
              _scores = evaluate(_nlp_model.tokenizer, _textClassifier, _testText, _testCats)```
